### PR TITLE
Make tensorflow optional in requirements and installation scripts

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

### Summary
Makes `tensorflow`, `tensorflow-datasets`, `tensorflow-text`, `Jetstream` and `seqio` optional dependencies in MaxText (Jetstream pulls tensorflow through Jetstream -> seqio_tensorflow-text -> tensorflow). Core pre-training lock files are now TensorFlow-free, while users requiring legacy TFDS pipelines or JetStream serving can install optional dependencies via `--with-tf`.

b/555209521

### Changes

* **Requirements & Lock Files**:
  * Removed TensorFlow, TFDS, TF-Text, and SeqIO from base requirements and generated lock files (`tpu-requirements.txt`, `cuda12-requirements.txt`, `decoupled-requirements.txt`).
  * Added `src/dependencies/extra_deps/tf_requirements.txt` and `jetstream_github_deps.txt` for optional installation.
* **Installation & Setup**:
  * Added `--with-tf` flag (and `WITH_TF` env var) to `install_pre_train_extra_deps.py` to install optional TensorFlow and JetStream dependencies.
  * Added `TF` option to `src/dependencies/scripts/setup.sh` and `docker_build_dependency_image.sh` (defaults to `false`).
  * Added `ARG TF=true` to Dockerfiles because nightly image workflow does not pass "TF" in build arg yet, will keep TF in nightly images until we build separated workflow for TF and no-TF images.
* **JetStream Stubs & Tests**:
  * Updated `gcloud_stub.py` to return stubs when JetStream or its dependencies are absent, avoiding import crashes.
  * Added skip guards to JetStream-dependent tests (`inference_microbenchmark_smoke_test.py`, `offline_engine_test.py`). The following tests are skipped in the no-TF environment. Will follow up with a dedicated jetstream CI workflow for these tests similar to the vllm post-training CI workflow.
```
    * `tests/integration/smoke/inference_microbenchmark_smoke_test.py`
    * `tests/unit/offline_engine_test.py`
    * `tests/integration/decode_tests.py`
    * `tests/integration/generate_param_only_checkpoint_test.py`
    * `tests/integration/maxengine_test.py`
    * `tests/unit/maxengine_nnx_test.py`
    * `tests/unit/deepseek_decode_consistency_test.py`
    * `tfds_c4_mlperf_stream_chunking_test.py`
    * `tfds_data_processing_test.py`
```   
  * Added automatic JetStream checks to 42 end-to-end test scripts calling `maxtext.inference.decode` to install `--with-tf` if absent.
* **Documentation**: Updated `install_maxtext.md` and `data_input_tfds.md` with `--with-tf` instructions.

# Tests
* github CI test pass (The failure in `tests/unit/mmap_data_processing_test.py::GrainMmapNpyEvalConfigTest` is at head and unrelated to this PR)
* Generated environments following this [instruction](https://github.com/AI-Hypercomputer/maxtext/blob/main/docs/install_maxtext.md#from-source) for cases with or without TF. Confirmed it works as expected.
```
uv pip install -e .[tpu] --resolution=lowest
install_tpu_pre_train_extra_deps  # tested with and without --with-tf
```
* Tested generating tf and no-tf maxtext images, confirm reduced image size (local uncompressed size 5.95 GB -> 3.91 GB, 35% reduction)
```
bash src/dependencies/scripts/docker_build_dependency_image.sh DEVICE=tpu MODE=stable TF=true LOCAL_IMAGE_NAME=maxtext_tf
bash src/dependencies/scripts/docker_build_dependency_image.sh DEVICE=tpu MODE=stable TF=false LOCAL_IMAGE_NAME=maxtext_no_tf
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/dependencies/extra_deps/tf_requirements.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+tensorflow==2.20.0
+tensorflow-datasets
+tensorflow-text==2.20.1
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/dependencies/extra_deps/tf_requirements.txt:3]: seqio is added

File: src/dependencies/scripts/install_pre_train_extra_deps.py [hunk 2]
```diff
@@ -70,6 +81,36 @@ def main():
...
+      print("Failed to install TensorFlow dependencies.")
+      print(f"Command '{' '.join(e.cmd)}' returned non-zero exit status {e.returncode}.")
+      print("--- Stderr ---")
+      print(e.stderr)
+      print("--- Stdout ---")
+      print(e.stdout)
+      sys.exit(e.returncode)
+    except (OSError, FileNotFoundError) as e:
+      print(f"An OS-level error occurred while trying to run uv: {e}")
+      sys.exit(1)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/dependencies/scripts/install_pre_train_extra_deps.py:104]: Good suggestion, updated

File: src/maxtext/inference/maxengine/maxengine.py [hunk 3]
```diff
@@ -1885,10 +1885,10 @@ def get_tokenizer(self) -> Any:
     """
     token_params_is_stub = getattr(_token_params_ns, "_IS_STUB", False)
     engine_api_is_stub = getattr(engine_api, "_IS_STUB", False)
-    if is_decoupled() and (token_params_is_stub or engine_api_is_stub):
+    if token_params_is_stub or engine_api_is_stub:
```
Comment (aireenmei) [inline src/maxtext/inference/maxengine/maxengine.py:1888]: Clarification: `is_decoupled()` is redundant here, token_params_is_stub or engine_api_is_stub is sufficient.

File: src/dependencies/scripts/install_pre_train_extra_deps.py [hunk 4]
```diff
@@ -56,20 +67,66 @@ def main():
...
+      raise FileNotFoundError(f"TensorFlow dependencies file not found at {tf_deps_path}")
+
+    tf_deps_command = [
+        sys.executable,
+        "-m",
+        "uv",
+        "pip",
+        "install",
+        "-r",
+        str(tf_deps_path),
```
Comment (SurbhiJainUSC) [inline src/dependencies/scripts/install_pre_train_extra_deps.py:92]: Should we install with `--no-deps`?

File: src/dependencies/requirements/generated_requirements/tpu-requirements.txt
Comment (SurbhiJainUSC) [inline src/dependencies/requirements/generated_requirements/tpu-requirements.txt:1]: Can you also update post-train requirements: https://github.com/AI-Hypercomputer/maxtext/blob/main/src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt?

File: src/dependencies/extra_deps/pre_train_github_deps.txt [hunk 5]
```diff
@@ -1,2 +1 @@
-google-jetstream @ https://github.com/AI-Hypercomputer/JetStream/archive/29329e8e73820993f77cfc8efe34eb2a73f5de98.zip
```
Comment (SurbhiJainUSC) [inline src/dependencies/extra_deps/pre_train_github_deps.txt:1]: Why are we creating separate txt file for Jetstream?

**Review by bvandermoon — COMMENTED** [review]
LGTM outside of @SurbhiJainUSC's comments

File: src/dependencies/extra_deps/pre_train_github_deps.txt (hunk 5 above)
Comment (aireenmei) [reply src/dependencies/extra_deps/pre_train_github_deps.txt:1]: I now deleted the separate txt file for jetstream and add jetstream to tf_requirements.txt

File: src/dependencies/requirements/generated_requirements/tpu-requirements.txt
Comment (aireenmei) [reply src/dependencies/requirements/generated_requirements/tpu-requirements.txt:1]: Done. Note that tensorflow_datasets is required by tunix, so it's still in the generated post_train_requirements. 

Additional changes made during the update:
* Added some version lower bounds to tpu_post_train_overrides.txt to resolve the warning messages and speed up the updating run of generate_requirements.sh
* Moving the override block in `generate_requirements.sh` before the post-train uv add step so that [tool.uv] override-dependencies from tpu_post_train_overrides.txt is present in pyproject.toml when resolving vllm and tpu-inference dependencies.

File: src/dependencies/scripts/install_pre_train_extra_deps.py (hunk 4 above)
Comment (aireenmei) [reply src/dependencies/scripts/install_pre_train_extra_deps.py:92]: We don't use `--no-deps` here because `tpu-requirements.txt` does not include sub-dependencies such as clu, dm-tree, pyglove, etc.

File: src/dependencies/extra_deps/pre_train_github_deps.txt (hunk 5 above)
Comment (SurbhiJainUSC) [reply src/dependencies/extra_deps/pre_train_github_deps.txt:1]: This is not related to TF right? Can we keep jetstream in pre_train_github_deps.txt only?

File: src/dependencies/scripts/install_pre_train_extra_deps.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply src/dependencies/scripts/install_pre_train_extra_deps.py:92]: We should always install other deps using --no-deps to maintain reproducibility in MaxText. I would suggest to add any transitive deps explicity in tf_requirements.txt.

cc: @bvandermoon

File: src/dependencies/extra_deps/pre_train_github_deps.txt (hunk 5 above)
Comment (aireenmei) [reply src/dependencies/extra_deps/pre_train_github_deps.txt:1]: Jetstream depends on TF (Jetstream -> seqio_tensorflow-text -> tensorflow)), so it needs to be moved out of pre_train_github_deps.txt

File: src/dependencies/scripts/install_pre_train_extra_deps.py (hunk 4 above)
Comment (aireenmei) [reply src/dependencies/scripts/install_pre_train_extra_deps.py:92]: I see. Now updated with `--no-deps` and includes all transitive deps in tf_requirements.txt. Also added a script generate_tf_requirements.py and a section in update_dependencies.md with instructions to update tf_requirements.txt

**Comment by aireenmei** [comment]
cc @snehalv2002 That the newly added `tfds_c4_mlperf_stream_chunking_test.py` will be skipped in the default github CI runner where no tf is present. Will create a dedicated CI workflow for tf tests (see the "JetStream Stubs & Tests" section in PR description).

**Review by bvandermoon — APPROVED** [review]

**Review by SurbhiJainUSC — APPROVED** [review]
